libcxx/test/lit.local.cfg

```
@@ -29,7 +29,13 @@ def appendToSubstitution(substitutions, key, value):
     return [(k, v + " " + value) if k == key else (k, v) for (k, v) in substitutions]
 
 
-std = getSubstitution("%{cxx_std}", config)
+# External users of the LLVM test suite might not have set the substitutions.
+std = ""
+try:
+    std = getSubstitution("%{cxx_std}", config)
+except:
+    pass
```

```suggestion
try:
    std = getSubstitution("%{cxx_std}", config)
except:
    std = ""
```
Seems like the more canonical way to me.

---

This substitution definitely should be provided by whoever runs the test suite -- it's defined in `params.py`. Is there something I misunderstand here?

The rest of the changes look fine, but this one seems like guarding against something we don't support and should never happen.

---

`cxx_std` is only ever used by libc++-specific tests, so I'm not sure it makes sense to require it to be provided.

---

But still, my point is that folks need to be running the test suite via the normal entry points which includes `params.py`, and I don't think it makes sense to try to support anything else. Otherwise we're bound to break them on a regular basis and that's not good for anyone. @StephanTLavavej Are you folks going through our normal entry point for Lit or do you still maintain your own custom Lit test format?

---

I don't have the best understanding of how we interface with Lit (i.e. how to recognize the "normal entry point"), but it's the stuff in https://github.com/microsoft/STL/tree/46843b3bf320c6070ff63e978936bcf898e5a2bc/tests/utils . We invoke our `stl-lit.py` generated from https://github.com/microsoft/STL/blob/46843b3bf320c6070ff63e978936bcf898e5a2bc/tests/utils/stl-lit/stl-lit.in and that eventually uses our `params.py` at https://github.com/microsoft/STL/blob/46843b3bf320c6070ff63e978936bcf898e5a2bc/tests/utils/stl/test/params.py . If there's something we should add there or to another file, I would be happy to do so.

(I think the raw string regex changes in this PR are independently needed, though.)

---

Oh, hmmm yeah that's quite a setup you folks have there. Normally, the intended way to use the test suite is to provide only this config file: https://github.com/llvm/llvm-project/blob/main/libcxx/test/configs/ibm-libc%2B%2B-shared.cfg.in (for example).

You set up the flags that are required for your stdlib, compiler, platform, etc, and then everything else should follow from that. I'm not sure why you're doing something significantly different here but I'd like to understand (and perhaps we can change a few things so that you folks can drop your custom format). As it stands, IIUC you basically use the contents of our test suite but almost none of our "Lit setup", and I think supporting that reliably is pretty much a lost battle.

The raw string regexes are definitely a non-controversial improvement. This change here I'm not like "over my dead body" against, but I think it does beg the question of how the test suite is actually being consumed and we should figure out the answer to that (and agree on what we want to support).

---

We have files like https://github.com/microsoft/STL/blob/46843b3bf320c6070ff63e978936bcf898e5a2bc/tests/std/lit.site.cfg.in . It sounds like we might need to add `config.substitutions.append(("%{cxx_std}", "cxx26"))` or maybe `lit_config.BLAH`?

> I'm not sure why you're doing something significantly different here

Me neither - @CaseyCarter might know more though.

---

@philnik777 Your suggestion changes the scope of `std`, which means the code no longer works. (In general I agree that your suggestion is the typical way to write it.)

@ldionne I was not sure we required all substitutions. AFAIK they are not documented and changes to them not mentioned in the release notes. I'll undo the changes to this file.